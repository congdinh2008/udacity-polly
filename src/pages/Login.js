import { useState } from "react";
import { FaPoll } from "react-icons/fa";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { loginUser } from "../actions/authedUser";
import { login } from "../apis/api";
const Login = () => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const navigate = useNavigate();

  const dispatch = useDispatch();

  const handleSubmit = (e) => {
    e.preventDefault();

    const user = { username, password };
    login(user)
      .then((res) => {
        if (res.success) {
          dispatch(loginUser(res.user));
          navigate("/");
        } else {
          setErrorMessage(res.message);
        }
      })
      .catch((err) => {
        setErrorMessage(err.message);
      });
  };

  return (
    <section className="form-section login-section d-flex align-items-center">
      <div className="form-container login-container text-center">
        <div className="form-header">
          <h1>Udacity Polly</h1>
          <FaPoll />
        </div>
        <div className="form login-form">
          <form onSubmit={handleSubmit}>
            <div className="form-group d-flex flex-column">
              <label htmlFor="username" className="form-control-label">
                Username
              </label>
              <input
                type="text"
                id="username"
                value={username}
                onChange={(event) => setUsername(event.target.value)}
              />
            </div>
            <div className="form-group d-flex flex-column">
              <label htmlFor="password" className="form-control-label">
                Password
              </label>
              <input
                type="password"
                id="password"
                value={password}
                onChange={(event) => setPassword(event.target.value)}
              />
            </div>
            <button className="btn btn-submit" type="submit">
              Login
            </button>
          </form>
          <span className={`error-message ${errorMessage ? "show" : ""}`}>
            {errorMessage}
          </span>
        </div>
      </div>
    </section>
  );
};

export default Login;
