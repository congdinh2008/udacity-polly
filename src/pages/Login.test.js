import { fireEvent, screen, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import { createStore } from "redux";
import configureMockStore from "redux-mock-store";
import Login from "./Login";
import reducer from "../reducers";
import middleware from "../middleware";

const mockStore = configureMockStore();

const users = {
  sarahedo: {
    id: "sarahedo",
    password: "password123",
    name: "Sarah Edo",
    avatarURL: "https://i.imgur.com/d8BSuEu.png",
    answers: {
      "8xf0y6ziyjabvozdd253nd": "optionOne",
      "6ni6ok3ym7mf1p33lnez": "optionOne",
      am8ehyc8byjqgar0jgpub9: "optionTwo",
      loxhs1bqm25b708cmbf3g: "optionTwo",
    },
    questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
  },
  tylermcginnis: {
    id: "tylermcginnis",
    password: "abc321",
    name: "Tyler McGinnis",
    avatarURL: "https://i.imgur.com/1gLpnj1.png",
    answers: {
      vthrdm985a262al8qx3do: "optionOne",
      xj352vofupe1dqz9emx13r: "optionTwo",
    },
    questions: ["loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do"],
  },
  mtsamis: {
    id: "mtsamis",
    password: "xyz123",
    name: "Mike Tsamis",
    avatarURL: "https://i.imgur.com/UI83gyB.png",
    answers: {
      xj352vofupe1dqz9emx13r: "optionOne",
      vthrdm985a262al8qx3do: "optionTwo",
      "6ni6ok3ym7mf1p33lnez": "optionOne",
    },
    questions: ["6ni6ok3ym7mf1p33lnez", "xj352vofupe1dqz9emx13r"],
  },
  zoshikanlu: {
    id: "zoshikanlu",
    password: "pass246",
    name: "Zenobia Oshikanlu",
    avatarURL: "https://i.imgur.com/b0cK6zx.png",
    answers: {
      xj352vofupe1dqz9emx13r: "optionOne",
    },
    questions: [],
  },
};

describe("Login", () => {
  it("will match snapshot", () => {
    const store = createStore(reducer, middleware);
    const view = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    expect(view).toMatchSnapshot();
  });

  it("select user to fill input", () => {
    const store = mockStore({
      users: users,
    });
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    const cards = screen.getAllByTestId("card-user");

    cards.forEach((card) => {
      fireEvent.click(card);

      const usernameInput = screen.getByTestId("username");
      const passwordInput = screen.getByTestId("password");

      setTimeout(() => {
        expect(usernameInput.value).toBe(card.key);
        expect(passwordInput.value).toBe(
          users.find((u) => u.id === card.key).password
        );
      }, 500);
    });
  });

  it("Login error with wrong username", () => {
    const store = mockStore({
      users: users,
    });
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    var userInput = screen.getByTestId("username");
    var passInput = screen.getByTestId("password");
    fireEvent.change(userInput, {
      target: { value: "wrongUserName" },
    });
    fireEvent.change(passInput, { target: { value: "password123" } });
    var submitButton = screen.getByTestId("btn-submit");
    fireEvent.click(submitButton);
    setTimeout(() => {
      expect(
        screen.getByText("Invalid username or password")
      ).toBeInTheDocument();
    }, 500);
  });

  it("login error with wrong password", () => {
    const store = mockStore({
      users: users,
    });
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    var userInput = screen.getByTestId("username");
    var passInput = screen.getByTestId("password");
    fireEvent.change(userInput, {
      target: { value: "sarahedo" },
    });
    fireEvent.change(passInput, { target: { value: "wrongPassword" } });
    var submitButton = screen.getByTestId("btn-submit");
    fireEvent.click(submitButton);
    setTimeout(() => {
      expect(
        screen.getByText("Invalid username or password")
      ).toBeInTheDocument();
    }, 500);
  });

  it("login successful", async () => {
    const store = mockStore({
      users: users,
    });

    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    var userInput = screen.getByTestId("username");
    var passInput = screen.getByTestId("password");

    fireEvent.change(userInput, {
      target: { value: "sarahedo" },
    });

    fireEvent.change(passInput, { target: { value: "password123" } });

    var submitButton = screen.getByTestId("btn-submit");

    fireEvent.click(submitButton);

    store.dispatch({ type: "LOGIN", payload: { user: users["sarahedo"] } });

    expect(store.getActions()).toEqual([
      { type: "LOGIN", payload: { user: users["sarahedo"] } },
    ]);
  });
});
