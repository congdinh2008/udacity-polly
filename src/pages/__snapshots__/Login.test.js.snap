// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="form-section login-section d-flex align-items-center"
      >
        <div
          class="form-container login-container text-center"
        >
          <div
            class="form-header"
          >
            <h1>
              Udacity Polly
            </h1>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM160 368c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16V240c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v128zm96 0c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16V144c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v224zm96 0c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16v-64c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v64z"
              />
            </svg>
          </div>
          <div
            class="select-user d-flex justify-content-center"
          />
          <div
            class="form login-form"
          >
            <form>
              <div
                class="form-group d-flex flex-column"
              >
                <label
                  class="form-control-label"
                  for="username"
                >
                  Username
                </label>
                <input
                  data-testid="username"
                  id="username"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group d-flex flex-column"
              >
                <label
                  class="form-control-label"
                  for="password"
                >
                  Password
                </label>
                <input
                  data-testid="password"
                  id="password"
                  type="password"
                  value=""
                />
              </div>
              <button
                class="btn btn-submit"
                data-testid="btn-submit"
                type="submit"
              >
                Login
              </button>
            </form>
            <span
              class="error-message "
            />
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="form-section login-section d-flex align-items-center"
    >
      <div
        class="form-container login-container text-center"
      >
        <div
          class="form-header"
        >
          <h1>
            Udacity Polly
          </h1>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM160 368c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16V240c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v128zm96 0c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16V144c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v224zm96 0c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16v-64c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v64z"
            />
          </svg>
        </div>
        <div
          class="select-user d-flex justify-content-center"
        />
        <div
          class="form login-form"
        >
          <form>
            <div
              class="form-group d-flex flex-column"
            >
              <label
                class="form-control-label"
                for="username"
              >
                Username
              </label>
              <input
                data-testid="username"
                id="username"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group d-flex flex-column"
            >
              <label
                class="form-control-label"
                for="password"
              >
                Password
              </label>
              <input
                data-testid="password"
                id="password"
                type="password"
                value=""
              />
            </div>
            <button
              class="btn btn-submit"
              data-testid="btn-submit"
              type="submit"
            >
              Login
            </button>
          </form>
          <span
            class="error-message "
          />
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
